CoreScriptsRoot/ServerCoreScripts/ClientChat/ChannelsBar.lua

```diff
@@ -0,0 +1,329 @@
+	return BaseFrame, ScrollerFrame, PageLeftButton, PageRightButton, LeaveConfirmationFrame, LeaveConfirmationNotice
+end
+
+
+function methods:UpdateMessagePostedInChannel(channelName)
+	local tab = self:GetChannelTab(channelName)
+	if (tab) then
+		tab:UpdateMessagePostedInChannel()
+	else
+		warn("ChannelsTab '" .. channelName .. "' does not exist!")
+	end
+end
+		
+function methods:AddChannelTab(channelName)
+	if (self:GetChannelTab(channelName)) then
+		error("Channel tab '" .. channelName .. "'already exists!")
+	end
+
+	local tab = moduleChannelsTab.new(channelName)
+	tab.GuiObject.Parent = self.ScrollerFrame
+	self.ChannelTabs[channelName:lower()] = tab
+
+	self.NumTabs = self.NumTabs + 1
+	self:OrganizeChannelTabs()
+
+	self.BackgroundTweener:RegisterTweenObjectProperty(tab.BackgroundTweener, "Transparency")
+	self.TextTweener:RegisterTweenObjectProperty(tab.TextTweener, "Transparency")
+
+	--// Although this feature is pretty much ready, it needs some UI design still.
+	local enableRightClickToLeaveChannel = false
+
+	if (enableRightClickToLeaveChannel) then
+		tab.NameTag.MouseButton2Click:connect(function()
+			self.LeaveConfirmationNotice.Text = string.format("Leave channel %s?", tab.ChannelName)
+			self.LeaveConfirmationFrame.LeaveTarget.Value = tab.ChannelName
+			self.LeaveConfirmationFrame:TweenPosition(UDim2.new(0, 0, 0, 0), Enum.EasingDirection.In, Enum.EasingStyle.Quad, 0.2, true)
+		end)
+	end
+
+	return tab
+end
+
+function methods:RemoveChannelTab(channelName)
+	if (not self:GetChannelTab(channelName)) then
+		error("Channel tab '" .. channelName .. "'does not exist!")
+	end
+
+	local indexName = channelName:lower() 
+	self.ChannelTabs[indexName]:Destroy()
+	self.ChannelTabs[indexName] = nil
+
+	self.NumTabs = self.NumTabs - 1
+	self:OrganizeChannelTabs()
+end
+
+function methods:GetChannelTab(channelName)
+	return self.ChannelTabs[channelName:lower()]
+end
+
+function methods:OrganizeChannelTabs()
+	local order = {}
+
+	table.insert(order, self:GetChannelTab("All"))
+	table.insert(order, self:GetChannelTab("System"))
+
+	for tabIndexName, tab in pairs(self.ChannelTabs) do
+		if (tab.ChannelName ~= "All" and tab.ChannelName ~= "System") then
+			table.insert(order, tab)
+		end
+	end
+
+	for index, tab in pairs(order) do
+		tab.GuiObject.Position = UDim2.new((index - 1) * 0.25, 0, 0, 0)
+	end
+
+	--// This does the effect of dynamic tab resizing when a player 
+	--// is in less than 4 channels. It was mathematically easier to
+	--// think about and do rather than resizing the actual tabs 
+	--// themselves.
+	local xSize = math.max(1, 4/self.NumTabs)
+	self.ScrollerFrame.Size = UDim2.new(xSize, 0, 1, 0)
+
+	self:ScrollChannelsFrame(0)
+end
+
+function methods:ScrollChannelsFrame(dir)
+	if (self.ScrollChannelsFrameLock) then return end
+	self.ScrollChannelsFrameLock = true
+
+	local newPageNum = self.CurPageNum + dir
+	if (newPageNum < 0) then
+		newPageNum = 0
+	elseif (newPageNum > 0 and newPageNum + 4 > self.NumTabs) then
```

Comment:
whats with the number 4? why do we want to do something different if 4 tabs? If this is necessary we should define at top of script, instead of using magic number everywhere